File: src/lib/updater.ts
// Lightweight update checker. Hits the GitHub Releases API once per
// app launch, compares the newest tag to the running version, and
// returns a descriptor for the footer toast to render. No auto-update
// — the user pilots the actual upgrade.
//
// GitHub's REST API is CORS-friendly and the unauthenticated rate
// limit is 60 req/hr/IP, which is plenty for a once-per-launch check.
// The request runs from the renderer (no Rust involvement) so there's
// no new dependency or backend surface.
//
// The repo is hardcoded here — if Baudrun ever forks or the user
// runs a mirror, they can edit this one constant.

const REPO = "packetThrower/Baudrun";

export type AvailableUpdate = {
  /** Version string without the leading "v" — e.g. `"1.0.0"` or `"1.1.0-beta.2"`. */
  version: string;
  /** Full tag name — e.g. `"v1.0.0"`. */
  tag: string;
  /** True if the release has a `-pre` suffix (alpha / beta / rc). */
  prerelease: boolean;
  /** Browser-facing URL for the release notes. */
  url: string;
  /** The current running version, carried through for UI display. */
  currentVersion: string;
};

type Parsed = {
  major: number;
  minor: number;
  patch: number;
  /** Pre-release identifier (the part after the `-`), or `null` for stable. */
  pre: string | null;
};

function parseSemver(raw: string): Parsed | null {
  const cleaned = raw.trim().replace(/^v/i, "");
  const m = cleaned.match(/^(\d+)\.(\d+)\.(\d+)(?:-(.+))?$/);
  if (!m) return null;
  return {
    major: Number(m[1]),
    minor: Number(m[2]),
    patch: Number(m[3]),
    pre: m[4] ?? null,
  };
}

/**
 * Classic SemVer precedence: compare numeric parts in order; a version
 * WITHOUT a pre-release suffix is always greater than the same numeric
 * core WITH one (1.0.0 > 1.0.0-rc.2). Within pre-releases, compare
 * lexicographically after splitting on `.` — enough for the alpha.1 <
 * alpha.2 < beta.1 ordering Baudrun's release workflow uses.
 *
 * Returns positive if `a > b`, negative if `a < b`, 0 if equal.
 */
export function compareSemver(a: string, b: string): number {
  const pa = parseSemver(a);
  const pb = parseSemver(b);
  if (!pa && !pb) return 0;
  if (!pa) return -1;
  if (!pb) return 1;
  if (pa.major !== pb.major) return pa.major - pb.major;
  if (pa.minor !== pb.minor) return pa.minor - pb.minor;
  if (pa.patch !== pb.patch) return pa.patch - pb.patch;
  if (pa.pre === null && pb.pre === null) return 0;
  if (pa.pre === null) return 1;
  if (pb.pre === null) return -1;
  return comparePreReleaseIds(pa.pre, pb.pre);
}

function comparePreReleaseIds(a: string, b: string): number {
  const as = a.split(".");
  const bs = b.split(".");
  const n = Math.max(as.length, bs.length);
  for (let i = 0; i < n; i++) {
    const ai = as[i];
    const bi = bs[i];
    if (ai === undefined) return -1; // shorter comes first (1.0.0-alpha < 1.0.0-alpha.1)
    if (bi === undefined) return 1;
    const an = Number(ai);
    const bn = Number(bi);
    if (!Number.isNaN(an) && !Number.isNaN(bn)) {
      if (an !== bn) return an - bn;
    } else {
      const cmp = ai.localeCompare(bi);
      if (cmp !== 0) return cmp;
    }
  }
  return 0;
}

type Release = {
  tag_name: string;
  html_url: string;
  prerelease: boolean;
  draft: boolean;
};

/** Hard cap on a single GitHub API response. A real /releases/latest
 *  payload runs ~5 KB and a /releases?per_page=10 list runs ~50 KB.
 *  100 KB is well above the ceiling, well below the renderer-DoS
 *  threshold a hostile redirect could otherwise leverage. */
const MAX_RESPONSE_BYTES = 100 * 1024;
const FETCH_TIMEOUT_MS = 10_000;

async function fetchReleases(includePrereleases: boolean): Promise<Release[]> {
  // `/releases/latest` returns a single JSON object and skips pre-releases + drafts.
  // `/releases?per_page=10` returns an array newest-first, including pre-releases.
  // Fetching 10 covers typical release cadence with headroom.
  const url = includePrereleases
    ? `https://api.github.com/repos/${REPO}/releases?per_page=10`
    : `https://api.github.com/repos/${REPO}/releases/latest`;

  // AbortController gives us a hard deadline AND a way to short-circuit
  // a slow / hung fetch — important on flaky networks where the
  // updater check would otherwise spin until the OS-level TCP timeout.
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);

  let resp: Response;
  try {
    resp = await fetch(url, {
      signal: controller.signal,
      headers: {
        Accept: "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
      },
    });
  } finally {
    clearTimeout(timer);
  }

  if (!resp.ok) throw new Error(`GitHub API ${resp.status}`);

  // Fast path: trust the Content-Length header when present. GitHub
  // sets it on every JSON response and an attacker who controls the
  // body but not the headers (e.g. a TLS MITM that can't forge certs
  // — i.e. nobody on a properly-secured connection) doesn't matter
  // here. Belt-and-suspenders is the byte-count loop below.
  const declaredSize = Number(resp.headers.get("content-length"));
  if (Number.isFinite(declaredSize) && declaredSize > MAX_RESPONSE_BYTES) {
    throw new Error(
      `GitHub API response declares ${declaredSize} bytes, capped at ${MAX_RESPONSE_BYTES}`,
    );
  }

  // Read incrementally so a server that lies about Content-Length (or
  // omits it on a chunked response) can't blow past the cap.
  const reader = resp.body?.getReader();
  if (!reader) throw new Error("GitHub API response had no body");
  const chunks: Uint8Array[] = [];
  let total = 0;
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    if (!value) continue;
    total += value.byteLength;
    if (total > MAX_RESPONSE_BYTES) {
      reader.cancel();
      throw new Error(
        `GitHub API response exceeded ${MAX_RESPONSE_BYTES} bytes`,
      );
    }
    chunks.push(value);
  }
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    merged.set(c, offset);
    offset += c.byteLength;
  }
  const body = new TextDecoder().decode(merged);
  const data = JSON.parse(body);
  if (Array.isArray(data)) return data as Release[];
  return [data as Release];
}

/**
 * Query GitHub for the newest release eligible under the given policy
 * and return an {@link AvailableUpdate} if it's ahead of `currentVersion`.
 * Returns `null` when there's nothing newer, when the request fails,
 * or when parsing falls over — callers treat it as "no update".
 */
export async function checkForUpdate(
  currentVersion: string,
  includePrereleases: boolean,
): Promise<AvailableUpdate | null> {
  try {
    const releases = await fetchReleases(includePrereleases);
    const eligible = releases
      .filter((r) => !r.draft)
      .filter((r) => includePrereleases || !r.prerelease);
    if (eligible.length === 0) return null;

    // Sort newest-first by semver so `/releases?per_page=10` order is
    // normalized even if GitHub's default sort drifts or a maintainer
    // re-tags an older release.
    eligible.sort((a, b) =>
      compareSemver(b.tag_name, a.tag_name),
    );
    const newest = eligible[0];
    const trimmed = newest.tag_name.replace(/^v/i, "");

    if (compareSemver(trimmed, currentVersion) <= 0) return null;

    return {
      version: trimmed,
      tag: newest.tag_name,
      prerelease: newest.prerelease,
      url: newest.html_url,
      currentVersion,
    };
  } catch (err) {
    console.warn("update check failed:", err);
    return null;
  }
}

File: src/updater.rs
//! Boot-time update check against the project's GitHub Releases.
//!
//! Posture by design: **detection only**. We never download a
//! replacement bundle or rewrite anything on disk. The check
//! discovers the latest tag, compares it against
//! `CARGO_PKG_VERSION`, and if a newer release exists publishes
//! a [`UpdateState`] global that the chrome reads to paint amber
//! indicators (Settings rail's `Updates` row + sidebar gear icon
//! in `app_view::sidebar_header`). The user opens the Updates
//! pane in Settings to see the version + release-notes link and
//! click through to the Releases page in their browser.
//!
//! Why no auto-install:
//!   * macOS code-signing + notarization haven't shipped yet.
//!     Auto-replacing a running unsigned bundle would trip
//!     Gatekeeper on every launch.
//!   * The "swap the live binary on quit" dance (Sparkle-style
//!     relauncher helper) is real work that's hard to verify on
//!     all three platforms.
//!   * Detection-only keeps the user in control — they pick
//!     whether and when to install.
//!
//! The fetch is sync (`ureq`), runs on a single
//! `gpui::BackgroundExecutor` task at boot. Worst case the user
//! is offline and the future resolves to `None` — no UI on the
//! amber-indicator path, exactly as if no update existed.

use std::time::Duration;

use serde::Deserialize;

/// Live state of the boot-time update check. Installed as a
/// gpui `Global` so any render path can read it cheaply (a
/// `.clone()` of a small Option). `None` means "no check has
/// completed yet" OR "the check ran and we're already on the
/// newest release" — render code treats both identically (no
/// amber dot painted).
#[derive(Debug, Clone, Default)]
pub struct UpdateState {
    /// `Some` when a newer release is available. Fields are the
    /// pieces of [`Release`] the chrome actually reads.
    pub available: Option<UpdateAvailable>,
}

impl gpui::Global for UpdateState {}

/// Available-update payload exposed to the chrome.
#[derive(Debug, Clone)]
pub struct UpdateAvailable {
    /// Bare tag string (no leading `v`), e.g. `"0.9.8"` or
    /// `"0.9.8-beta.2"`. Compared against `dismissed_update_version`
    /// for re-prompting suppression.
    pub version: String,
    /// Browser URL for the GitHub Releases page entry. Opened
    /// when the user clicks the Updates pane's "View release"
    /// button.
    pub html_url: String,
    /// Optional release-notes body (Markdown). Truncated for
    /// in-app display — the full notes are on the html_url.
    pub notes: String,
}

/// GitHub Releases API response shape — we only deserialize the
/// fields we actually use. `#[serde(default)]` on optional
/// strings so partial responses don't fail the whole parse.
#[derive(Debug, Deserialize)]
struct Release {
    #[serde(default)]
    tag_name: String,
    #[serde(default)]
    html_url: String,
    #[serde(default)]
    body: String,
    #[serde(default)]
    draft: bool,
    #[serde(default)]
    prerelease: bool,
}

/// Query the project's GitHub Releases API for the newest release
/// that the user's `include_prerelease` preference allows.
///
/// Returns `Ok(None)` when no release is newer than
/// `current_version`, when the network call fails, or when the
/// response shape doesn't parse. Returns `Err` only for
/// programmer-error paths (the current version string didn't
/// parse as semver — should never happen since it comes from
/// `env!("CARGO_PKG_VERSION")` which Cargo validates at compile
/// time).
///
/// Blocking — call from a `BackgroundExecutor` task, never from
/// the render thread.
pub fn check_for_update(
    current_version: &str,
    include_prerelease: bool,
) -> Result<Option<UpdateAvailable>, semver::Error> {
    let current = semver::Version::parse(current_version)?;
    let releases = match fetch_releases(include_prerelease) {
        Some(r) => r,
        None => return Ok(None),
    };
    let newest = releases
        .into_iter()
        // Drafts are unpublished and shouldn't show up via the
        // public API, but filter defensively in case GitHub's
        // pagination ever leaks one.
        .filter(|r| !r.draft)
        // Hide pre-releases unless the user opted in.
        .filter(|r| include_prerelease || !r.prerelease)
        // Parse each tag's semver; tags that don't parse get
        // silently skipped. Most repo tags are `vX.Y.Z[-pre]`
        // — strip the leading `v`.
        .filter_map(|r| {
            let bare = r.tag_name.strip_prefix('v').unwrap_or(&r.tag_name);
            let ver = semver::Version::parse(bare).ok()?;
            Some((ver, r))
        })
        // Pick the maximum by version, NOT by API-order, because
        // GitHub sorts by created_at and a backported patch on
        // an older line could land newest-by-date.
        .max_by(|(a, _), (b, _)| a.cmp(b));
    let Some((newest_ver, release)) = newest else {
        return Ok(None);
    };
    if newest_ver <= current {
        return Ok(None);
    }
    Ok(Some(UpdateAvailable {
        version: newest_ver.to_string(),
        html_url: release.html_url,
        notes: release.body,
    }))
}

/// Hit the GitHub Releases API. Returns `None` on any error —
/// the caller treats that the same as "no update found", so the
/// user never sees a "check failed" diagnostic. Errors land in
/// the standard log sink at `info!` level (not `warn!` /
/// `error!` — a transient offline state isn't a problem to
/// surface).
fn fetch_releases(include_prerelease: bool) -> Option<Vec<Release>> {
    // Stable-only path uses `/releases/latest` (single object).
    // Pre-release path needs `/releases` (paginated; first page
    // is 30 entries which is plenty since we only want the
    // newest). Wrapping the singleton in a Vec keeps the
    // downstream filter chain uniform.
    let url = if include_prerelease {
        "https://api.github.com/repos/packetThrower/Baudrun/releases?per_page=30"
    } else {
        "https://api.github.com/repos/packetThrower/Baudrun/releases/latest"
    };

    // 5s connect + 5s read — long enough that a slow Wi-Fi hop
    // doesn't false-negative, short enough that a captive-portal
    // hang doesn't keep the worker thread blocked forever.
    let agent = ureq::AgentBuilder::new()
        .timeout_connect(Duration::from_secs(5))
        .timeout_read(Duration::from_secs(5))
        .user_agent(concat!(
            "Baudrun/",
            env!("CARGO_PKG_VERSION"),
            " (https://github.com/packetThrower/Baudrun)"
        ))
        .build();
    let response = match agent.get(url).call() {
        Ok(r) => r,
        Err(err) => {
            log::info!("update check: HTTP {err}");
            return None;
        }
    };
    let parsed = if include_prerelease {
        response.into_json::<Vec<Release>>()
    } else {
        response.into_json::<Release>().map(|r| vec![r])
    };
    match parsed {
        Ok(v) => Some(v),
        Err(err) => {
            log::info!("update check: parse failed: {err}");
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn current_version_parses_as_semver() {
        // Sanity that the build-time-substituted version string
        // is parseable — `env!("CARGO_PKG_VERSION")` is what the
        // boot-time check feeds into `check_for_update`.
        let raw = env!("CARGO_PKG_VERSION");
        semver::Version::parse(raw).expect("CARGO_PKG_VERSION must be valid semver");
    }
}
